src/main/java/com/artipie/asto/s3/EstimatedContentCompliment.java

```diff
@@ -0,0 +1,164 @@
+     *
+     * @param original Original content.
+     * @param limit Content reading limit.
+     */
+    EstimatedContentCompliment(final Content original, final long limit) {
+        this.original = original;
+        this.limit = limit;
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param original Original content.
+     */
+    EstimatedContentCompliment(final Content original) {
+        this(original, Long.MAX_VALUE);
+    }
+
+    /**
+     * Initialize future of Content.
+     *
+     * @return The future.
+     */
+    public CompletionStage<Content> estimate() {
+        final CompletableFuture<Content> res;
+        if (this.original.size().isPresent()) {
+            res = CompletableFuture.completedFuture(this.original);
+        } else {
+            res = this.readUntilLimit();
+        }
+        return res;
+    }
+
+    /**
+     * Read until limit.
+     *
+     * @return The future.
+     */
+    private CompletableFuture<Content> readUntilLimit() {
+        final Path temp;
+        try {
+            temp = Files.createTempFile(
+                S3Storage.class.getSimpleName(),
+                ".upload.tmp"
+            );
+        } catch (final IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+        return new File(temp)
+            .write(this.original)
+            .thenCompose(
+                nothing ->
+                    Flowable.fromPublisher(
+                        new File(temp).content()
+                    )
+                        .map(Buffer::remaining)
+                        .scanWith(() -> 0L, Long::sum)
+                        .takeUntil(total -> total >= this.limit)
+                        .lastOrError()
+                        .to(SingleInterop.get())
+                        .toCompletableFuture()
+            )
+            .thenApply(
+                last -> {
+                    final Optional<Long> size;
+                    if (last >= this.limit) {
+                        size = Optional.empty();
+                    } else {
+                        size = Optional.of(last);
+                    }
+                    return size;
+                }
+            ).thenApply(
+                sizeOpt -> {
+                    final Flowable<ByteBuffer> data = Flowable.fromPublisher(
+                        new File(temp).content()
+                    ).doAfterTerminate(
+                        () -> Files.delete(temp)
+                    );
+                    return sizeOpt
+                        .<Content>map(size -> new Content.From(size, data))
+                        .orElse(new Content.From(data));
+                }
+            )
+            .handle(
+                (value, throwable) -> {
+                    final Content result;
+                    if (throwable == null) {
+                        result = value;
+                    } else {
+                        try {
+                            Files.delete(temp);
```

Comment:
@andreoss here: we can remove tmp file in a single place without checking if error is not `null`

Reply:
@g4s8 I removed `handle` completely, and moved cleanup to `Flowable` hooks